_deploy/r/gnoswap/gns/halving.gno

```diff
@@ -30,67 +51,48 @@ var (
+	endTimestamp int64 // (exactly) after 12 years, timestamp which gns emission will be ended
+)
 
-var halvingYearAmount = make(map[int64]uint64)     // year => mintAmount
-var halvingYearMintAmount = make(map[int64]uint64) // year => (actual) mintAmount
-var halvingYearAccuAmount = make(map[int64]uint64) // year => accuAmount
-func init() {
-	// yearly mint amount
-	halvingYearAmount[1] = 18_750_000_000_000 * 12 // 225000000000000
-	halvingYearAmount[2] = 18_750_000_000_000 * 12 // 225000000000000
-	halvingYearAmount[3] = 9_375_000_000_000 * 12  // 112500000000000
-	halvingYearAmount[4] = 9_375_000_000_000 * 12  // 112500000000000
-	halvingYearAmount[5] = 4_687_500_000_000 * 12  // 56250000000000
-	halvingYearAmount[6] = 4_687_500_000_000 * 12  // 56250000000000
-	halvingYearAmount[7] = 2_343_750_000_000 * 12  // 28125000000000
-	halvingYearAmount[8] = 2_343_750_000_000 * 12  // 28125000000000
-	halvingYearAmount[9] = 1_171_875_000_000 * 12  // 14062500000000
-	halvingYearAmount[10] = 1_171_875_000_000 * 12 // 14062500000000
-	halvingYearAmount[11] = 1_171_875_000_000 * 12 // 14062500000000
-	halvingYearAmount[12] = 1_171_875_000_000 * 12 // 14062500000000
-
-	// yearly accumulated mint amount
-	halvingYearAccuAmount[1] = halvingYearAmount[1]
-	halvingYearAccuAmount[2] = halvingYearAccuAmount[1] + halvingYearAmount[2]
-	halvingYearAccuAmount[3] = halvingYearAccuAmount[2] + halvingYearAmount[3]
-	halvingYearAccuAmount[4] = halvingYearAccuAmount[3] + halvingYearAmount[4]
-	halvingYearAccuAmount[5] = halvingYearAccuAmount[4] + halvingYearAmount[5]
-	halvingYearAccuAmount[6] = halvingYearAccuAmount[5] + halvingYearAmount[6]
-	halvingYearAccuAmount[7] = halvingYearAccuAmount[6] + halvingYearAmount[7]
-	halvingYearAccuAmount[8] = halvingYearAccuAmount[7] + halvingYearAmount[8]
-	halvingYearAccuAmount[9] = halvingYearAccuAmount[8] + halvingYearAmount[9]
-	halvingYearAccuAmount[10] = halvingYearAccuAmount[9] + halvingYearAmount[10]
-	halvingYearAccuAmount[11] = halvingYearAccuAmount[10] + halvingYearAmount[11]
-	halvingYearAccuAmount[12] = halvingYearAccuAmount[11] + halvingYearAmount[12]
-}
+var (
+	halvingYearBlock     = make([]int64, HALVING_END_YEAR) // start block of each halving year
+	halvingYearTimestamp = make([]int64, HALVING_END_YEAR) // start timestamp of each halving year
 
-var amountPerBlockPerHalvingYear = make(map[int64]uint64) // year => reward per block
+	halvingYearMaxAmount         = make([]uint64, HALVING_END_YEAR)
+	halvingYearMintAmount        = make([]uint64, HALVING_END_YEAR)
+	halvingYearAccuAmount        = make([]uint64, HALVING_END_YEAR)
+	amountPerBlockPerHalvingYear = make([]uint64, HALVING_END_YEAR)
+)
 
 func init() {
-	height := std.GetHeight()
-
-	startHeight = height
+	startHeight = std.GetHeight()
 	startTimestamp = time.Now().Unix()
 
-	initialized = true
+	for year := HALVING_START_YEAR; year <= HALVING_END_YEAR; year++ {
```

Comment:
It would be better to extract the loop's internal logic. Because the `init` function cannot be tested, but we can verify the extracted function.

Reply:
I don't quite get it, can you enlighten me with more detail? (small peace of suggestion  would be perfect)

Reply:
theres no other details for this line. I don't prefer hard coded value